crates/amaru-ledger/src/state/diff_bind.rs

```diff
@@ -32,14 +32,28 @@ impl<K: Ord, J, V> Default for DiffBind<K, J, V> {
     }
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Key is already registered")]
+    KeyAlreadyRegistered,
+    #[error("Key is already unregistered")]
+    KeyAlreadyUnregistered,
+}
+
 impl<K: Ord, J: Clone, V> DiffBind<K, J, V> {
-    pub fn register(&mut self, k: K, v: V, j: Option<J>) {
+    pub fn register(&mut self, k: K, v: V, j: Option<J>) -> Result<(), Error> {
+        if self.registered.contains_key(&k) {
+            return Err(Error::KeyAlreadyUnregistered);
+        }
         self.unregistered.remove(&k);
         self.registered.insert(k, (j, Some(v)));
+        Ok(())
     }
 
-    pub fn bind(&mut self, k: K, j: Option<J>) {
-        assert!(!self.unregistered.contains(&k));
```

Comment:
We could use `clippy` to prevent assert usage vi this [lint](https://rust-lang.github.io/rust-clippy/master/index.html#/disallowed_macros) unfortunately it can't be configured to be usable in tests too. See https://github.com/rust-lang/rust-clippy/issues/13790